### PR TITLE
Create bank repository and current repository.

### PR DESCRIPTION
**Pull request by aleska351**
Added method retrieve, create and delete.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -14,7 +14,10 @@ repositories {
 
 dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-web'
-    runtimeOnly 'org.postgresql:postgresql'
+    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+    implementation 'org.postgresql:postgresql:42.2.18'
+
+    //runtimeOnly 'org.postgresql:postgresql'
```
Comment (Andriireva) [inline build.gradle]: delete unused dependency

File: src/main/java/com/goncharenko/currencyexchangerate/CurrencyexchangerateApplication.java [hunk 2]
```diff
@@ -1,19 +1,22 @@
...
-        String nameApp = (String) context.getBean("nameFromRoot");
-        System.out.println(nameApp);
-        System.out.println(context.getBean("currentTime"));
-        System.out.println(context.getBean("nameFromString"));
+       SpringApplication.run(CurrencyexchangerateApplication.class, args);
+       LOGGER.info("Application run");
+//        String nameApp = (String) context.getBean("nameFromRoot");
+//        System.out.println(nameApp);
+//        System.out.println(context.getBean("currentTime"));
+//        System.out.println(context.getBean("nameFromString"));
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/CurrencyexchangerateApplication.java]: delete unused code

File: src/main/java/com/goncharenko/currencyexchangerate/dao/BankRepository.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Optional;
+
+@Component
+public interface BankRepository {
+    Optional<Bank> retrieveById(Long id);
+
+    Optional<List<Bank>> retrieveAll();
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dao/BankRepository.java]: Make it return LIst<Bank> without Optional

File: src/main/java/com/goncharenko/currencyexchangerate/dao/CurrencyRepository.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+import java.util.Optional;
+
+public interface CurrencyRepository {
+
+    Optional<Currency> retrieveById(Long id);
+
+    Optional<List<Currency>> retrieveAll();
+
+    Optional<List<Currency>> retrieveAllCurrenciesByBankId(Long bankId);
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dao/CurrencyRepository.java]: Make it return List<Currency> without Optional

File: src/main/java/com/goncharenko/currencyexchangerate/dao/impl/BankRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    private final JdbcTemplate jdbcTemplate;
+    private final CurrencyRepository currencyRepository;
+    private final BankRowMapper bankMapper;
+
+    @Autowired
+    public BankRepositoryImpl(JdbcTemplate jdbcTemplate, CurrencyRepository currencyRepository, BankRowMapper bankMapper) {
+        this.currencyRepository = currencyRepository;
+        this.jdbcTemplate = jdbcTemplate;
+        this.bankMapper = bankMapper;
+        LOGGER.info("================BankRepository constructor is called===========");
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dao/impl/BankRepositoryImpl.java]: it is better to make it on debug level

File: src/main/java/com/goncharenko/currencyexchangerate/dao/impl/BankRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * This method retrieves bank by id
+     *
+     * @param id bank id
+     * @return current bank or Optional.empty() if bank with given id not found
+     */
+    @Override
+    public Optional<Bank> retrieveById(Long id) {
+        LOGGER.info("retrieve bank with id {} ", id);
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dao/impl/BankRepositoryImpl.java]: put it on debug level

File: src/main/java/com/goncharenko/currencyexchangerate/dao/impl/BankRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+     * This method deletes bank with given id in table
+     *
+     * @param id bank ID which need to delete
+     */
+    @Override
+    public void delete(Long id) {
+
+        jdbcTemplate.update("DELETE from banks  WHERE id = ?", id);
+        // delete all bank's currencies after bank deletion
+        currencyRepository.deleteAllByBankId(id);
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dao/impl/BankRepositoryImpl.java]: It is not needed on Repository level

File: src/main/java/com/goncharenko/currencyexchangerate/dao/impl/BankRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import java.sql.PreparedStatement;
+import java.util.List;
+import java.util.Optional;
+
+
+@Component
+public class BankRepositoryImpl implements BankRepository {
+    private static final Logger LOGGER = LoggerFactory.getLogger(BankRepositoryImpl.class);
+    private final JdbcTemplate jdbcTemplate;
+    private final CurrencyRepository currencyRepository;
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dao/impl/BankRepositoryImpl.java]: It is bad practice to use another repository in current repository. Repository is a simple and "dump" class that do a single action, it does not serve as a logic class

File: src/main/java/com/goncharenko/currencyexchangerate/dao/impl/CurrencyRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    private final JdbcTemplate jdbcTemplate;
+    private final CurrencyRowMapper currencyRowMapper;
+
+
+    public CurrencyRepositoryImpl(JdbcTemplate jdbcTemplate, CurrencyRowMapper currencyRowMapper) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.currencyRowMapper = currencyRowMapper;
+        LOGGER.info("================CurrencyRepository constructor is called===========");
+
+
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dao/impl/CurrencyRepositoryImpl.java]: remove empty rows

File: src/main/java/com/goncharenko/currencyexchangerate/dao/impl/CurrencyRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.List;
+import java.util.Optional;
+
+
+@Component
+public class CurrencyRepositoryImpl implements CurrencyRepository {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CurrencyRepositoryImpl.class);
+    private final JdbcTemplate jdbcTemplate;
+    private final CurrencyRowMapper currencyRowMapper;
+
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dao/impl/CurrencyRepositoryImpl.java:25]: remove 1 row

File: src/main/java/com/goncharenko/currencyexchangerate/domain/Bank.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+    public enum Type {
+
+        GLOBAL(),
+        LOCAL();
+
+        Type() {
+
+        }
+    }
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/domain/Bank.java]: It is better to move it to a separate file

File: src/main/java/com/goncharenko/currencyexchangerate/mapper/CurrencyRowMapper.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
+package com.goncharenko.currencyexchangerate.mapper;
+
+import com.goncharenko.currencyexchangerate.domain.Currency;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.stereotype.Component;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+@Component
+public class CurrencyRowMapper implements RowMapper<Currency> {
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/mapper/CurrencyRowMapper.java:11]: it is a good point that you made it as bean.

**Review by Andriireva — CHANGES_REQUESTED** [review]

**Review by aleska351 — COMMENTED** [review]
Review code on your recommendation

File: src/main/java/com/goncharenko/currencyexchangerate/dto/CurrencyDTO.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
+package com.goncharenko.currencyexchangerate.dto;
+
+import com.goncharenko.currencyexchangerate.domain.Currency;
+import org.springframework.stereotype.Component;
+
+import java.util.Objects;
+
+@Component
```
Comment (Andriireva) [inline src/main/java/com/goncharenko/currencyexchangerate/dto/CurrencyDTO.java:8]: Зачем нужен компонент? Вроде как инжекта нигде нет на этот класс.

**Pull request closed without merging**